server/src/infra/ldap_handler.rs

```diff
@@ -120,15 +120,54 @@ fn get_user_attribute(user: &User, attribute: &str, dn: &str) -> Result<Option<V
     }))
 }
 
+fn expand_attribute_wildcards(
+    attributes: &[String],
+    all_attribute_keys: &[&str],
+    operational_attribute_keys: &[&str],
+) -> Vec<String> {
+    let mut attributes_out = attributes.to_owned();
+
+    if let Some(all_idx) = attributes_out.iter().position(|x| x == "*") {
+        debug!(r#"Expanding *"#);
+        // Splice in all non-operational attributes
+        attributes_out.splice(
```

Comment:
`splice` is not what you're looking for here. Maybe `extend_from_slice`?
I'm trying to think of ways to avoid copying too much, though, so how about this:
`make_ldap_search_user_result_entry` creates a `Vec<&str>` from the `&[String]`, and here we extend it with the 'static str since they're hardcoded.

Reply:
i think splice, or at least something that removes the original element, because we certainly don't want to keep the '*' and '+'

Reply:
Consider replacing vecs withs sets 

> nitnelave — Today at 23:19
> Right, splice might be what we need indeed
> The constant can live next to the function
> Make sure that you test with multiple *
> strazto — Today at 23:20
> oh, true...
> nitnelave — Today at 23:20
> Actually, maybe having a set rather than vec would make sense
> strazto — Today at 23:20
> also true
> that's a bigger restructure though
> is supplying multiple * valid user input?
> / supplying multiple of any attribute key
> nitnelave — Today at 23:23
> There's valid input, and what people will try :) we should support it
> strazto — Today at 23:24
> as in, clean it up, or at least give a better error message
> nitnelave — Today at 23:24
> I think with the Set<&str> idea it shouldn't be too much effort
> strazto — Today at 23:25
> my experience is that when it comes to configuring ldap is that everyone ive ever seen try it just does random sh*t until it works
> nitnelave — Today at 23:25
> Now that I think about it, having string rather than str is probably okay, with the small string optimization
> And it'll probably be simpler to write
> strazto — Today at 23:25
> i think you've lost me
> nitnelave — Today at 23:26
> :D
> In your expand function, replace the input from &[String] to &mut Set<String>
> strazto — Today at 23:27
> gotcha, so just don't worry about changing String -> &str
> nitnelave — Today at 23:27
> Hmm, though apparently rust doesn't have that optimization
> I'll have to read more about it, but write with String to start with
> Then we'll see if we want to replace it
> strazto — Today at 23:28
> sounds good, can be optimzied if it's actually causing a problem later
> something something premature optimization
> nitnelave — Today at 23:28
> It's most likely premature optimization, but I don't care, it's my hobby project :D
> strazto — Today at 23:28
> that FASTEST hobby ever
> nitnelave — Today at 23:29
> Haha

Reply:
I think HashSets are premature optimization.

A valid query is `attrs: ['*', '+']` or `attrs: [ 'lldap schema attributes']` .

These are both bounded by O(1)

except in the edge case a user gives a tonne of duplicate attributes (  `attrs: ['*', '+', '*', '+', '*', '+', '*', '+', '*', '+', '*', '+', '*', '+', '*', '+', ]`
This behaviour is handled, but performance will be linearly bounded by number of duplicate attrs, which doesn't matter because they don't deserve good performance.